pcf/particle/aws/sns/sns_topic.py

```diff
@@ -0,0 +1,213 @@
+        """
+        if self._arn:
+            try:
+                current_definition = self.client.get_topic_attributes(
+                    TopicArn=self._arn
+                )
+                current_definition["Name"] = self.topic_name
+                self.current_state_definition = current_definition
+                return current_definition
+            except ClientError:
+                return None
+        else:
+            return None
+
+    def get_subscription(self):
+        """
+        Uses boto calls to get the Subscription Attributes for the Topic
+
+        Returns:
+            subscription definition if the topic exists, otherwise None
+        """
+        if self.subscription_arn:
+            try:
+                subscription_definition = self.client.get_subscription_attributes(
+                    SubscriptionArn=self.subscription_arn_arn
+                )
+                return subscription_definition
+            except ClientError:
+                return None
+        else:
+            return None
+
+    def sync_state(self):
+        """
+        Uses get_current_definition to determine whether the topic exists or not and sets the state
+
+        Returns:
+            void
+        """
+        # get the current definition. if it exists, running; if missing, terminated.
+        if self.get_current_definition():
+            self.state = State.running
+        else:
+            self.state = State.terminated
+
+    def is_state_definition_equivalent(self):
+        """
+        Compared the desired state and current state definition
+
+        Returns:
+            bool
+        """
+        self.sync_state()
+        # use filters to remove any extra information
+        if self.custom_config.get("Attributes"):
+            self.desired_state_definition["Attributes"] = self.custom_config.get("Attributes")
+
+        self.current_state_definition = pcf_util.param_filter(self.current_state_definition, SNSTopic.DEFINITION_FILTER)
+        self.desired_state_definition = pcf_util.param_filter(self.desired_state_definition, SNSTopic.DEFINITION_FILTER)
+        if "Attributes" in self.desired_state_definition.keys():
+            self.desired_state_definition["Attributes"] = pcf_util.param_filter(
+                self.desired_state_definition.get("Attributes"), SNSTopic.START_ATTR)
+            # only compare attributes specified in desired, ignore all else
+            self.current_state_definition["Attributes"] = pcf_util.param_filter(
+                self.current_state_definition.get("Attributes"), self.desired_state_definition.get("Attributes").keys())
+
+        diff_dict = pcf_util.diff_dict(self.current_state_definition, self.desired_state_definition)
+        return diff_dict == {}
+
+    def _update(self):
+        """
+        Updates any changed attributes
+
+        Returns:
+            void
+        """
+        if not self.is_state_definition_equivalent():
```

Comment:
dont need to call this since update is called after is_state_definition_equivalent in core particle